File: src/render/procedural.js
import { CELL_WIDTH, CELL_HEIGHT, fillRect, setPixel } from "./grid.js";

// Box drawing, block elements and sextants are drawn procedurally (not from
// font bitmaps) so they tile seamlessly across cells — the whole point of
// mosaic/ASCII-art content. Scope, per the plan: light and heavy single-
// weight lines/corners/tees/cross only; dashed, double, rounded and
// diagonal box-drawing variants are out of scope and fall through undrawn.

const LIGHT = 1;
const HEAVY = 2;
const THICKNESS = { [LIGHT]: 2, [HEAVY]: 4 };

// { up, down, left, right } edge weights (0 = no edge).
const BOX_CHARS = {
    0x2500: { left: LIGHT, right: LIGHT }, // ─
    0x2502: { up: LIGHT, down: LIGHT }, // │
    0x250c: { down: LIGHT, right: LIGHT }, // ┌
    0x2510: { down: LIGHT, left: LIGHT }, // ┐
    0x2514: { up: LIGHT, right: LIGHT }, // └
    0x2518: { up: LIGHT, left: LIGHT }, // ┘
    0x251c: { up: LIGHT, down: LIGHT, right: LIGHT }, // ├
    0x2524: { up: LIGHT, down: LIGHT, left: LIGHT }, // ┤
    0x252c: { down: LIGHT, left: LIGHT, right: LIGHT }, // ┬
    0x2534: { up: LIGHT, left: LIGHT, right: LIGHT }, // ┴
    0x253c: { up: LIGHT, down: LIGHT, left: LIGHT, right: LIGHT }, // ┼
    0x2501: { left: HEAVY, right: HEAVY }, // ━
    0x2503: { up: HEAVY, down: HEAVY }, // ┃
    0x250f: { down: HEAVY, right: HEAVY }, // ┏
    0x2513: { down: HEAVY, left: HEAVY }, // ┓
    0x2517: { up: HEAVY, right: HEAVY }, // ┗
    0x251b: { up: HEAVY, left: HEAVY }, // ┛
    0x2523: { up: HEAVY, down: HEAVY, right: HEAVY }, // ┣
    0x252b: { up: HEAVY, down: HEAVY, left: HEAVY }, // ┫
    0x2533: { down: HEAVY, left: HEAVY, right: HEAVY }, // ┳
    0x253b: { up: HEAVY, left: HEAVY, right: HEAVY }, // ┻
    0x254b: { up: HEAVY, down: HEAVY, left: HEAVY, right: HEAVY }, // ╋
};

function drawBox(fb, originX, originY, codepoint, fgIndex) {
    const spec = BOX_CHARS[codepoint];
    if (!spec) return false;
    const cx = originX + CELL_WIDTH / 2;
    const cy = originY + CELL_HEIGHT / 2;
    if (spec.up) {
        const w = THICKNESS[spec.up];
        fillRect(fb, cx - w / 2, originY, w, cy - originY + w / 2, fgIndex);
    }
    if (spec.down) {
        const w = THICKNESS[spec.down];
        fillRect(fb, cx - w / 2, cy - w / 2, w, originY + CELL_HEIGHT - (cy - w / 2), fgIndex);
    }
    if (spec.left) {
        const w = THICKNESS[spec.left];
        fillRect(fb, originX, cy - w / 2, cx - originX + w / 2, w, fgIndex);
    }
    if (spec.right) {
        const w = THICKNESS[spec.right];
        fillRect(fb, cx - w / 2, cy - w / 2, originX + CELL_WIDTH - (cx - w / 2), w, fgIndex);
    }
    return true;
}

const QUADRANT = {
    UL: { x: 0, y: 0 },
    UR: { x: CELL_WIDTH / 2, y: 0 },
    LL: { x: 0, y: CELL_HEIGHT / 2 },
    LR: { x: CELL_WIDTH / 2, y: CELL_HEIGHT / 2 },
};
const QUADRANT_W = CELL_WIDTH / 2;
const QUADRANT_H = CELL_HEIGHT / 2;

// Height eighths (24/8 = 3px, exact); width eighths (12/8 = 1.5px) are
// rounded since 12 doesn't divide 8 evenly — only the sextant grid (/2, /3)
// is guaranteed exact, per the plan's font/grid decision.
const eighthHeight = (n) => Math.round((CELL_HEIGHT * n) / 8);
const eighthWidth = (n) => Math.round((CELL_WIDTH * n) / 8);

function blockRects(codepoint) {
    switch (codepoint) {
        case 0x2580:
            return [{ x: 0, y: 0, w: CELL_WIDTH, h: CELL_HEIGHT / 2 }]; // upper half
        case 0x2584:
            return [{ x: 0, y: CELL_HEIGHT / 2, w: CELL_WIDTH, h: CELL_HEIGHT / 2 }]; // lower half
        case 0x2588:
            return [{ x: 0, y: 0, w: CELL_WIDTH, h: CELL_HEIGHT }]; // full block
        case 0x258c:
            return [{ x: 0, y: 0, w: CELL_WIDTH / 2, h: CELL_HEIGHT }]; // left half
        case 0x2590:
            return [{ x: CELL_WIDTH / 2, y: 0, w: CELL_WIDTH / 2, h: CELL_HEIGHT }]; // right half
        case 0x2581:
        case 0x2582:
        case 0x2583:
        case 0x2585:
        case 0x2586:
        case 0x2587: {
            const n = codepoint - 0x2580; // 1..7, one eighth to seven eighths
            const h = eighthHeight(n);
            return [{ x: 0, y: CELL_HEIGHT - h, w: CELL_WIDTH, h }];
        }
        case 0x2589:
        case 0x258a:
        case 0x258b:
        case 0x258d:
        case 0x258e:
        case 0x258f: {
            const n = 8 - (codepoint - 0x2588); // 2589->7 ... 258f->1
            const w = eighthWidth(n);
            return [{ x: 0, y: 0, w, h: CELL_HEIGHT }];
        }
        case 0x2594:
            return [{ x: 0, y: 0, w: CELL_WIDTH, h: eighthHeight(1) }]; // upper one eighth
        case 0x2595: {
            const w = eighthWidth(1);
            return [{ x: CELL_WIDTH - w, y: 0, w, h: CELL_HEIGHT }]; // right one eighth
        }
        case 0x2596:
            return [QUADRANT.LL];
        case 0x2597:
            return [QUADRANT.LR];
        case 0x2598:
            return [QUADRANT.UL];
        case 0x2599:
            return [QUADRANT.UL, QUADRANT.LL, QUADRANT.LR];
        case 0x259a:
            return [QUADRANT.UL, QUADRANT.LR];
        case 0x259b:
            return [QUADRANT.UL, QUADRANT.UR, QUADRANT.LL];
        case 0x259c:
            return [QUADRANT.UL, QUADRANT.UR, QUADRANT.LR];
        case 0x259d:
            return [QUADRANT.UR];
        case 0x259e:
            return [QUADRANT.UR, QUADRANT.LL];
        case 0x259f:
            return [QUADRANT.UR, QUADRANT.LL, QUADRANT.LR];
        default:
            return null;
    }
}

// Approximate stipple, not the real Unicode shade bitmap — deterministic and
// good enough for a teletext-style mosaic at this cell size.
function drawShade(fb, originX, originY, density, fgIndex) {
    for (let y = 0; y < CELL_HEIGHT; y++) {
        for (let x = 0; x < CELL_WIDTH; x++) {
            const on = density === 0.25 ? (x + y) % 4 === 0 : density === 0.5 ? (x + y) % 2 === 0 : (x + y) % 4 !== 0;
            if (on) setPixel(fb, originX + x, originY + y, fgIndex);
        }
    }
    return true;
}

function drawBlock(fb, originX, originY, codepoint, fgIndex) {
    if (codepoint === 0x2591) return drawShade(fb, originX, originY, 0.25, fgIndex);
    if (codepoint === 0x2592) return drawShade(fb, originX, originY, 0.5, fgIndex);
    if (codepoint === 0x2593) return drawShade(fb, originX, originY, 0.75, fgIndex);
    const rects = blockRects(codepoint);
    if (!rects) return false;
    for (const r of rects) {
        fillRect(fb, originX + r.x, originY + r.y, r.w ?? QUADRANT_W, r.h ?? QUADRANT_H, fgIndex);
    }
    return true;
}

// Sextant construction verified against the Unicode 17.0 code chart: masks
// 1-62 excluding 21 (left column, == U+258C) and 42 (right column, ==
// U+2590), ascending, mapped sequentially from U+1FB00 — lands exactly on
// U+1FB3B (60 codepoints).
export const SEXTANT_BASE = 0x1fb00;
export const SEXTANT_MASKS = buildSextantMasks();

function buildSextantMasks() {
    const masks = {};
    let codepoint = SEXTANT_BASE;
    for (let mask = 1; mask <= 62; mask++) {
        if (mask === 21 || mask === 42) continue;
        masks[codepoint] = mask;
        codepoint++;
    }
    return masks;
}

// bit -> sub-cell (2 cols x 3 rows), verified against masks 21 (bits 0,2,4 =
// left column = U+258C) and 42 (bits 1,3,5 = right column = U+2590).
const SEXTANT_CELLS = [
    { bit: 0, col: 0, row: 0 },
    { bit: 1, col: 1, row: 0 },
    { bit: 2, col: 0, row: 1 },
    { bit: 3, col: 1, row: 1 },
    { bit: 4, col: 0, row: 2 },
    { bit: 5, col: 1, row: 2 },
];

function drawSextant(fb, originX, originY, codepoint, fgIndex) {
    const mask = SEXTANT_MASKS[codepoint];
    if (mask === undefined) return false;
    const w = CELL_WIDTH / 2;
    const h = CELL_HEIGHT / 3;
    for (const cell of SEXTANT_CELLS) {
        if (mask & (1 << cell.bit)) {
            fillRect(fb, originX + cell.col * w, originY + cell.row * h, w, h, fgIndex);
        }
    }
    return true;
}

// Dispatches by codepoint range. Returns false (nothing drawn) for anything
// out of scope, so the caller can fall back to glyphs.js/blank.
export function drawProcedural(fb, originX, originY, codepoint, fgIndex) {
    if (codepoint >= 0x1fb00 && codepoint <= 0x1fb3b) return drawSextant(fb, originX, originY, codepoint, fgIndex);
    if (codepoint >= 0x2580 && codepoint <= 0x259f) return drawBlock(fb, originX, originY, codepoint, fgIndex);
    if (codepoint >= 0x2500 && codepoint <= 0x257f) return drawBox(fb, originX, originY, codepoint, fgIndex);
    return false;
}

File: sketches/videotext/src/render/procedural.cpp
// Box drawing, block elements and sextants are drawn as rectangles rather
// than font bitmaps so they tile seamlessly across cells, which is the whole
// point of mosaic and ASCII-art content. Light and heavy single-weight lines,
// corners, tees and crosses only; other box-drawing variants fall through.

#include "draw.h"

namespace vt {
namespace {

constexpr int HALF_W = CELL_WIDTH / 2;
constexpr int HALF_H = CELL_HEIGHT / 2;

// Edge weights: 0 none, 1 light (2 px), 2 heavy (4 px).
struct BoxSpec {
    char32_t codepoint;
    std::uint8_t up, down, left, right;
};

constexpr BoxSpec BOX_CHARS[] = {
    {0x2500, 0, 0, 1, 1},  // ─
    {0x2502, 1, 1, 0, 0},  // │
    {0x250C, 0, 1, 0, 1},  // ┌
    {0x2510, 0, 1, 1, 0},  // ┐
    {0x2514, 1, 0, 0, 1},  // └
    {0x2518, 1, 0, 1, 0},  // ┘
    {0x251C, 1, 1, 0, 1},  // ├
    {0x2524, 1, 1, 1, 0},  // ┤
    {0x252C, 0, 1, 1, 1},  // ┬
    {0x2534, 1, 0, 1, 1},  // ┴
    {0x253C, 1, 1, 1, 1},  // ┼
    {0x2501, 0, 0, 2, 2},  // ━
    {0x2503, 2, 2, 0, 0},  // ┃
    {0x250F, 0, 2, 0, 2},  // ┏
    {0x2513, 0, 2, 2, 0},  // ┓
    {0x2517, 2, 0, 0, 2},  // ┗
    {0x251B, 2, 0, 2, 0},  // ┛
    {0x2523, 2, 2, 0, 2},  // ┣
    {0x252B, 2, 2, 2, 0},  // ┫
    {0x2533, 0, 2, 2, 2},  // ┳
    {0x253B, 2, 0, 2, 2},  // ┻
    {0x254B, 2, 2, 2, 2},  // ╋
};

int thickness(std::uint8_t weight) {
    return weight == 2 ? 4 : 2;
}

bool drawBox(std::uint8_t* fb, int x, int y, char32_t codepoint, std::uint8_t color) {
    const BoxSpec* spec = nullptr;
    for (const BoxSpec& s : BOX_CHARS) {
        if (s.codepoint == codepoint) spec = &s;
    }
    if (spec == nullptr) return false;
    const int cx = x + HALF_W;
    const int cy = y + HALF_H;
    if (spec->up) {
        const int w = thickness(spec->up);
        fillRect(fb, cx - w / 2, y, w, cy - y + w / 2, color);
    }
    if (spec->down) {
        const int w = thickness(spec->down);
        fillRect(fb, cx - w / 2, cy - w / 2, w, y + CELL_HEIGHT - (cy - w / 2), color);
    }
    if (spec->left) {
        const int w = thickness(spec->left);
        fillRect(fb, x, cy - w / 2, cx - x + w / 2, w, color);
    }
    if (spec->right) {
        const int w = thickness(spec->right);
        fillRect(fb, cx - w / 2, cy - w / 2, x + CELL_WIDTH - (cx - w / 2), w, color);
    }
    return true;
}

// Height eighths are exact (3 px); width eighths (1.5 px) round half up,
// as Math.round did. Only the sextant grid (/2, /3) is guaranteed exact.
int eighthHeight(int n) {
    return CELL_HEIGHT * n / 8;
}

int eighthWidth(int n) {
    return (CELL_WIDTH * n + 4) / 8;
}

enum Quadrant : std::uint8_t { UL = 1, UR = 2, LL = 4, LR = 8 };

void fillQuadrants(std::uint8_t* fb, int x, int y, int quadrants, std::uint8_t color) {
    if (quadrants & UL) fillRect(fb, x, y, HALF_W, HALF_H, color);
    if (quadrants & UR) fillRect(fb, x + HALF_W, y, HALF_W, HALF_H, color);
    if (quadrants & LL) fillRect(fb, x, y + HALF_H, HALF_W, HALF_H, color);
    if (quadrants & LR) fillRect(fb, x + HALF_W, y + HALF_H, HALF_W, HALF_H, color);
}

// An approximate stipple, not the Unicode shade bitmaps. 25% and 75% use a
// 2x2 ordered (Bayer) tile so they read as an even dot grid instead of
// diagonal hatching; 50% is a plain checkerboard.
void drawShade(std::uint8_t* fb, int x, int y, int percent, std::uint8_t color) {
    constexpr int BAYER_2X2[2][2] = {{0, 2}, {3, 1}};
    for (int dy = 0; dy < CELL_HEIGHT; dy++) {
        for (int dx = 0; dx < CELL_WIDTH; dx++) {
            bool on;
            if (percent == 50) {
                on = (dx + dy) % 2 == 0;
            } else {
                const int threshold = percent == 25 ? 1 : 3;  // 1 of 4 vs 3 of 4 lit
                on = BAYER_2X2[dy % 2][dx % 2] < threshold;
            }
            if (on) setPixel(fb, x + dx, y + dy, color);
        }
    }
}

bool drawBlock(std::uint8_t* fb, int x, int y, char32_t codepoint, std::uint8_t color) {
    switch (codepoint) {
        case 0x2580:  // upper half
            fillRect(fb, x, y, CELL_WIDTH, HALF_H, color);
            return true;
        case 0x2584:  // lower half
            fillRect(fb, x, y + HALF_H, CELL_WIDTH, HALF_H, color);
            return true;
        case 0x2588:  // full block
            fillRect(fb, x, y, CELL_WIDTH, CELL_HEIGHT, color);
            return true;
        case 0x258C:  // left half
            fillRect(fb, x, y, HALF_W, CELL_HEIGHT, color);
            return true;
        case 0x2590:  // right half
            fillRect(fb, x + HALF_W, y, HALF_W, CELL_HEIGHT, color);
            return true;
        case 0x2581:
        case 0x2582:
        case 0x2583:
        case 0x2585:
        case 0x2586:
        case 0x2587: {  // lower one to seven eighths
            const int h = eighthHeight(static_cast<int>(codepoint - 0x2580));
            fillRect(fb, x, y + CELL_HEIGHT - h, CELL_WIDTH, h, color);
            return true;
        }
        case 0x2589:
        case 0x258A:
        case 0x258B:
        case 0x258D:
        case 0x258E:
        case 0x258F: {  // left seven eighths down to one eighth
            const int w = eighthWidth(8 - static_cast<int>(codepoint - 0x2588));
            fillRect(fb, x, y, w, CELL_HEIGHT, color);
            return true;
        }
        case 0x2591:
            drawShade(fb, x, y, 25, color);
            return true;
        case 0x2592:
            drawShade(fb, x, y, 50, color);
            return true;
        case 0x2593:
            drawShade(fb, x, y, 75, color);
            return true;
        case 0x2594:  // upper one eighth
            fillRect(fb, x, y, CELL_WIDTH, eighthHeight(1), color);
            return true;
        case 0x2595: {  // right one eighth
            const int w = eighthWidth(1);
            fillRect(fb, x + CELL_WIDTH - w, y, w, CELL_HEIGHT, color);
            return true;
        }
        case 0x2596:
            fillQuadrants(fb, x, y, LL, color);
            return true;
        case 0x2597:
            fillQuadrants(fb, x, y, LR, color);
            return true;
        case 0x2598:
            fillQuadrants(fb, x, y, UL, color);
            return true;
        case 0x2599:
            fillQuadrants(fb, x, y, UL | LL | LR, color);
            return true;
        case 0x259A:
            fillQuadrants(fb, x, y, UL | LR, color);
            return true;
        case 0x259B:
            fillQuadrants(fb, x, y, UL | UR | LL, color);
            return true;
        case 0x259C:
            fillQuadrants(fb, x, y, UL | UR | LR, color);
            return true;
        case 0x259D:
            fillQuadrants(fb, x, y, UR, color);
            return true;
        case 0x259E:
            fillQuadrants(fb, x, y, UR | LL, color);
            return true;
        case 0x259F:
            fillQuadrants(fb, x, y, UR | LL | LR, color);
            return true;
        default:
            return false;
    }
}

// U+1FB00..U+1FB3B are the 2x3 masks 1-62 in ascending order, skipping 21
// (left column, already U+258C) and 42 (right column, already U+2590). Bits
// 0-5 run left to right, top to bottom.
bool drawSextant(std::uint8_t* fb, int x, int y, char32_t codepoint, std::uint8_t color) {
    if (codepoint < 0x1FB00 || codepoint > 0x1FB3B) return false;
    int mask = static_cast<int>(codepoint - 0x1FB00) + 1;
    if (mask >= 21) mask++;
    if (mask >= 42) mask++;
    const int w = CELL_WIDTH / 2;
    const int h = CELL_HEIGHT / 3;
    for (int bit = 0; bit < 6; bit++) {
        if (mask & (1 << bit)) fillRect(fb, x + (bit % 2) * w, y + (bit / 2) * h, w, h, color);
    }
    return true;
}

}  // namespace

bool isProceduralCodepoint(char32_t codepoint) {
    return (codepoint >= 0x1FB00 && codepoint <= 0x1FB3B) || (codepoint >= 0x2580 && codepoint <= 0x259F) ||
           (codepoint >= 0x2500 && codepoint <= 0x257F);
}

bool drawProcedural(std::uint8_t* fb, int x, int y, char32_t codepoint, std::uint8_t color) {
    if (codepoint >= 0x1FB00 && codepoint <= 0x1FB3B) return drawSextant(fb, x, y, codepoint, color);
    if (codepoint >= 0x2580 && codepoint <= 0x259F) return drawBlock(fb, x, y, codepoint, color);
    if (codepoint >= 0x2500 && codepoint <= 0x257F) return drawBox(fb, x, y, codepoint, color);
    return false;
}

}  // namespace vt
